File: frontend/src/api/usage.ts
import type { LlmUsageSample } from "../types";
import { buildQuery, readCachedGet, request } from "./client";

export interface LlmUsageResponse {
  items: LlmUsageSample[];
  total: number;
}

export async function fetchLlmUsage(limit = 72, includeTotal = true): Promise<LlmUsageResponse> {
  return request<LlmUsageResponse>(buildLlmUsagePath(limit, includeTotal));
}

export function readCachedLlmUsage(limit = 72, includeTotal = true): LlmUsageResponse | null {
  return readCachedGet<LlmUsageResponse>(buildLlmUsagePath(limit, includeTotal));
}

function buildLlmUsagePath(limit = 72, includeTotal = true): string {
  return `/api/llm-usage${buildQuery({ limit, include_total: includeTotal ? undefined : false })}`;
}

File: backend/app/api/usage.py
from typing import Annotated

import oracledb
from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.errors import oracle_http_exception
from app.core.security import require_admin_module
from app.repositories.usage import list_llm_usage
from app.schemas.usage import LlmUsageResponse


router = APIRouter(prefix="/llm-usage", tags=["llm-usage"], dependencies=[Depends(require_admin_module("usage"))])


@router.get("", response_model=LlmUsageResponse)
async def get_llm_usage(
    limit: Annotated[int, Query(ge=1, le=240)] = 72,
    include_total: bool = True,
) -> LlmUsageResponse:
    try:
        items, total = await list_llm_usage(limit=limit, include_total=include_total)
    except oracledb.Error as exc:
        raise oracle_http_exception(exc, "Oracle rejected the LLM usage query") from exc

    return LlmUsageResponse(items=items, total=total)
